src/System.IO.FileSystem/src/System/IO/DirectoryInfo.cs

```diff
@@ -61,17 +61,21 @@ public DirectoryInfo CreateSubdirectory(string path)
             if (Path.IsPathRooted(path))
                 throw new ArgumentException(SR.Arg_Path2IsRooted, nameof(path));
 
-            string fullPath = Path.GetFullPath(Path.Combine(FullPath, path));
+            string newPath = Path.GetFullPath(Path.Combine(FullPath, path));
 
-            if (fullPath.Length < FullPath.Length 
-                || (fullPath.Length > FullPath.Length && !PathInternal.IsDirectorySeparator(fullPath[FullPath.Length]))
-                || string.Compare(FullPath, 0, fullPath, 0, FullPath.Length, PathInternal.StringComparison) != 0)
+            ReadOnlySpan<char> trimmedNewPath = PathInternal.TrimEndingDirectorySeparator(newPath.AsSpan());
+            ReadOnlySpan<char> trimmedCurrentPath = PathInternal.TrimEndingDirectorySeparator(FullPath.AsSpan());
+
+            // We want to make sure the requested directory is actually under the subdirectory.
+            if (!trimmedNewPath.StartsWith(trimmedCurrentPath, PathInternal.StringComparison)
```

Comment:
I found it easier to grok if I flip it around, up to you
```C#
            // Make sure the requested directory is actually under the subdirectory.                         
            if (trimmedNewPath.StartsWith(trimmedCurrentPath, PathInternal.StringComparison) 
            {
                // Note we must also check the first character after the current path is a slash.
                // Otherwise passing "c:\foo" and "..\foobar", which would produce a new path of "c:\foobar", would be allowed.
                if ((trimmedNewPath.Length == trimmedCurrentPath.Length) || PathInternal.IsDirectorySeparator(newPath[trimmedCurrentPath.Length])
                {
            		FileSystem.CreateDirectory(newPath);
            		return new DirectoryInfo(newPath);
            	}
            }

            throw new ArgumentException(SR.Format(SR.Argument_InvalidSubPath, path, FullPath), nameof(path));
```

Reply:
> I found it easier to grok if I flip it around

I agree- it reduces the negatives. Updating.